onnxruntime/core/framework/mem_pattern_planner.h

```diff
@@ -31,7 +31,38 @@ class MemPatternPlanner {
  public:
   MemPatternPlanner() = default;
 
-  void TraceAllocation(int ml_value_idx, size_t size) {
+  // Returns true if there is an intersection between two time schedules.
+  // ASSUMES EACH TIME SCHEDULE IS SORTED. THIS IS VALIDATED AT THE END OF MEMORY PLANNING.
+  bool OverlappingTimeSchedules(const std::vector<size_t>& program_counter_start_1, const std::vector<size_t>& program_counter_end_1,
+                                const std::vector<size_t>& program_counter_start_2, const std::vector<size_t>& program_counter_end_2) {
+    ORT_ENFORCE(program_counter_start_1.size() > 0);
+    ORT_ENFORCE(program_counter_start_2.size() > 0);
+    ORT_ENFORCE(program_counter_start_1.size() == program_counter_end_1.size());
+    ORT_ENFORCE(program_counter_start_2.size() == program_counter_end_2.size());
+
+    size_t index_1 = 0;
+    size_t index_2 = 0;
+    bool overlap = false;
+    while ((index_1 < program_counter_start_1.size()) && (index_2 < program_counter_start_2.size())) {
+      if (program_counter_start_1[index_1] <= program_counter_start_2[index_2]) {
+        if (program_counter_end_1[index_1] >= program_counter_start_2[index_2]) {
+          overlap = true;
+          break;
```

Comment:
(nit):  just `return true;` for these and `return false;` outside the loop.

Reply:
I'll add this in a subsequent PR right after this.

Reply:
https://github.com/microsoft/onnxruntime/pull/5659